stash/src/service/MgxAirdropService.ts

```diff
@@ -0,0 +1,126 @@
+import MangataClient from '../connector/MangataNode.js'
+import { BN } from '@polkadot/util'
+import { ApiDecoration } from '@polkadot/api/types/index.js'
+import { BN_ZERO } from 'gasp-sdk'
+import { u128, u32 } from '@polkadot/types-codec'
+import { signatureVerify } from '@polkadot/util-crypto'
+
+const MGX_TOKEN_ID = '0'
+const AIRDROP_WEIGHT = 0.5
+
+export const getEligibilityAtBlockN = async (
+  block: string,
+  address: string
+) => {
+  const api = await (await MangataClient.api()).at(block)
+
+  const balances = await getBalances(api, address)
+  const issuance = await getTokenIssuance(api)
+
+  const mgxPools = await getMgxPools(api)
+  const poolMetadata = await getPoolMetadata(api, mgxPools)
+
+  const lpAssets = Array.from(balances.keys()).filter((k) =>
+    poolMetadata.has(k)
+  )
+
+  const mgxBalance = balances.get(MGX_TOKEN_ID)
+  const mgxWeight = mgxBalance.muln(AIRDROP_WEIGHT * 10).divn(10)
+  const lpMgxHoldings = lpAssets.map((lpId) => {
+    return calculateTokenHoldings(
+      poolMetadata.get(lpId),
+      balances.get(lpId),
+      issuance.get(lpId)
+    )
+  })
+
+  const lpMgxHoldingsSum = lpMgxHoldings.reduce(
+    (acc, weight) => acc.add(weight),
+    BN_ZERO
+  )
+  const lpWeight = lpMgxHoldingsSum.muln(AIRDROP_WEIGHT * 10).divn(10)
+
+  return {
+    weight: mgxWeight.add(lpWeight),
+    blockMgxHoldings: mgxBalance,
+    blockMgxLpHoldings: lpMgxHoldingsSum,
+    block,
+  }
+}
+
+export function verifySignature(
+  data: Record<'mangataXAddress' | 'ethereumAddress', string>,
+  signature: string
+) {
+  const message = `0x${Buffer.from(JSON.stringify(data)).toString('hex')}`
+  const { isValid } = signatureVerify(message, signature, data.mangataXAddress)
+
+  return isValid
+}
+
+function calculateTokenHoldings(
+  poolMetadata: Record<string, u128>,
+  balance: BN,
+  issuance: BN
+) {
+  const holdings = poolMetadata[MGX_TOKEN_ID].mul(balance).div(issuance).muln(2)
```

Comment:
will poolMetadata always have MGX_TOKEN_ID? or it can be null?

Reply:
`poolMetadata` only returns pools with MGX pairs so there will always be MGX in these pools